### DISCUSSION TITLE
invalidateQueries stops working if I navigate between Next.js routes

### DISCUSSION THREAD

**Discussion by detoner777** [TanStack/query#3037] [General] [answered]
We used ```queryClient.invalidateQueries(someQueryKey)``` to refetch data in queries  and its works well.
Project migrated to ```Next.js ``` and router changed from ```react-router``` to ```next-router``` code base stays maximum the same except for router changes.

Now if I navigate from one route to another and then go back, **queryClient.invalidateQueries(someQueryKey)** stops working, all other query calls work as it should useQueris and useMutations . 

What weird, when I navigate between the pages queries disappear from ```Query Dev Tools``` but continue to work, except of  ```queryClient.invalidateQueries()```. If I refresh the page ```Query Dev Tools```  starts to show me the correct queries and it starts working correctly until I navigate to another route and back. 

I'd really appreciate if anyone can help me with this issue! Let me know if any additional info is required.

**Comment by TkDodo** [comment, marked as answer]
You really need to show the code. My best guess is you create a new QueryClient on every page transition, which basically throws away the old cache.

**Reply by yordis** [reply]
@TkDodo do you recommend doing `React.useState` or `React.useMemo` of some sort?

**Reply by yordis** [reply]
The following seems to fix the issue, indeed, the previous one was creating a new client every time, so I assume that is why the query invalidations weren't working properly.

```tsx
export function useReactQueryClient() {
  const router = useRouter();
  const [queryClient] = React.useState(function () {
    return new QueryClient({
      defaultOptions: {
        queries: {
          onError(err) {
            // ....
          },
        },
      },
    });
  });

  return queryClient;
}
```

**Reply by TkDodo** [reply]
yes, your client has to be stable. However `useMemo` does not guarantee this. Have a read here: https://tkdodo.eu/blog/use-state-for-one-time-initializations

**Reply by yordis** [reply]
Roger! Thank you very much (fixing the example btw)
